### ISSUE TITLE
Uncaught TypeError: Illegal constructor Error

### ISSUE THREAD

**Issue by ayumihamsaki** [PrismJS/prism#1181]
Hi,

I am getting this error:

```
Uncaught TypeError: Illegal constructor
    at n (prism.js:1)
    at prism.js:1
    at prism.js:1
    at Array.forEach (<anonymous>)
    at Prism.plugins.toolbar.hook (prism.js:1)
    at Object.run (prism.js:1)
    at Object.highlightElement (prism.js:1)
    at HTMLDocument.highlightAll (prism.js:1)
```

It seems to be showing when I add the "**Copy to Clipboard Button**" package.

My browser I am getting the error is Google Chrome Version 62.0.3200.0 (Official Build)

Here is a screenshot of the error in dev tools:

![untitled](https://user-images.githubusercontent.com/17784082/29998966-f85d1852-902f-11e7-8207-c4a6d82c0f03.png)

For now I am just going to live with out adding this plugin. I thought I'd let you know anyway.

p.s. Normal setup works fine no problem. It only appears with this plugin addon.

**Comment by mAAdhaTTah** [comment]
Is [Clipboard.js](https://clipboardjs.com/) included on the page? Do you have a public, reproducible example?

**Comment by ayumihamsaki** [comment]
@mAAdhaTTah I don't have clipboard.js installed on that website. Also I have prism running in the backend of a custom cms so I can't show you the URL.

**Comment by mAAdhaTTah** [comment]
You need Clipboard.js installed. That's what handles the clipboard interactions.

**Comment by jakeNiemiec** [comment]
Same issue here

**Comment by jakeNiemiec** [comment]
This was also a problem with webpack.

The problem was the same: use clipboard/lib/clipboard.js instead of the dist. This seems to be an issue with how they transpile.

**Comment by knyy** [comment]
Any chance your browser is Chrome v61?

Chrome v61 added a window.Clipboard which is probably causing an namespace collision with copyToClipboard plugin.

**Comment by mAAdhaTTah** [comment]
Did it really? We check `window.Clipboard` to determine whether to inject/provide the CDN version. If Chrome has a `window.Clipboard` already, we can't check inject the correct one.

Poking at it in the console, it doesn't appear to... do anything. I can't `new` it, it has no static methods, although there are a few prototype methods, and I can't find any documentation on it. There's also a ticket open @ Clipboard.js to change the global name: https://github.com/zenorocha/clipboard.js/issues/468

Hoping Chrome will provide some docs on this feature so maybe we can figure out how to detect it properly or Clipboard.js changes the global name. I don't have a solution off the top though.

**Comment by js-digital** [comment]
Hi, 
The same issue I got in the latest version of Chrome.
[Version 61.0.3163.79 (Official Build) (64-bit)]

Screenshot from devTool
<img width="1440" alt="screen shot 2017-09-12 at 10 36 19 am" src="https://user-images.githubusercontent.com/11888586/30316159-fc7876a8-97ae-11e7-91e2-75b9060a1747.png">

**Comment by ayumihamsaki** [comment]
Interesting points was not aware of needing Clipboard.js installed.

Actually I am running Chrome V63 see here:

![as](https://user-images.githubusercontent.com/17784082/30335476-650e0324-97da-11e7-89cf-3dea8c1e850e.png)

**Comment by mAAdhaTTah** [comment]
@ayumihamsaki Anything after 61 will have this problem, as that's when the global `Clipboard` object was introduced. The plugin normally installs Clipboard.js at runtime for you, if you don't install it yourself, but with the introduction of `window.Clipboard` in Chrome 61, this check fails, as `window.Clipboard` already exists. It's just not the `Clipboard.js` instance we're expecting here.

**Comment by caraya** [comment]
I was having the same problem with Prism installed manually in a Wordpress theme. 

**Adding clipboard.js explicitly solved the issue in Chrome 61 and Canary (63.0.3213.0).**  

@mAAdhaTTah Chromestatus has an entry for [Asynchronous Clipboard API](https://www.chromestatus.com/features/5861289330999296). It also has a couple links to demos and docs.

The links point to an [explainer](https://github.com/garykac/clipboard/blob/master/clipboard.md) and the [proposal at wicg](https://discourse.wicg.io/t/proposal-modern-asynchronous-clipboard-api/1513)

Hope this helps

**Comment by mAAdhaTTah** [comment]
@caraya Thanks, that might be it, but Chrome 61 (or 63) doesn't have a `navigator.clipboard` property. I'm still not sure the cause.

Just so everyone knows, this can be resolved by including Clipboard.js yourself, either in a script tag or bundled into your JS. It's specifically the plugin's behavior to fallback to a CDN version if it doesn't find Clipboard.js that is failing.

**Comment by nunos7** [comment]
In my scenario
```
<script src="prism.js"></script>
<script src="prism-toolbar.min.js"></script>
<script src="copy-to-clipboard/prism-copy-to-clipboard.min.js"></script>
```

i just inverting the position of the copy-to-clipboard plugin to be the first one to be loaded, and problem solved!.
```
<script src="copy-to-clipboard/prism-copy-to-clipboard.min.js"></script>
<script src="prism.js"></script>
<script src="prism-toolbar.min.js"></script>
```

However loading the plugin before the main component seems a bit strange...

NS

**Comment by timhemendinger** [comment]
I added:

`$("<script />", { type: "text/javascript", src: "https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.5.8/clipboard.min.js" }).appendTo("head");` right after `Prism = function () {` in prism.js. 

It's essentially the same thing as adding the clipboard script to `<head>` but I wanted to keep it contained to the script itself.

**Comment by ldemailly** [comment]
@timhemendinger I tried that but didn't get it to work - do you have a diff ?

**Comment by rwebster85** [comment]
If you're using PrismJS in a custom wordpress plugin, you can enqueue "clipboard.min.js" before you enqueue "prismJS" like so...

`
wp_register_script( 'prismJS', plugins_url('/includes/prism.js',__FILE__ ));
wp_register_script( 'clipboardPrismJS', 'https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.5.8/clipboard.min.js' );
wp_enqueue_script('clipboardPrismJS');
wp_enqueue_script('prismJS');
`

**Comment by mAAdhaTTah** [comment]
So the `require` step doesn't work either because `window.Clipboard` was found initially. The only way this works is to load up the Clipboard first and assign the correct instance to `window.Clipboard` D:.

**Comment by mAAdhaTTah** [comment]
@Golmote Would it be a terrible hack to do something like this instead:

```js
/(native code)/.test(window.Clipboard.toString())
```

**Comment by mAAdhaTTah** [comment]
This version of the fix is here: #1206.
